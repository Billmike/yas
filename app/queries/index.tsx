import { gql } from '@apollo/client';

export const GET_CHARACTERS = gql`
  query GetCharacters($page: Int, $name: String!) {
    characters(page: $page, filter: { name: $name }) {
      info {
        next
      }
      results {
        name
        image
        id
      }
    }
  }
`;

export const GET_CHARACTER = gql`
  query Character($id: ID!) {
    character(id: $id) {
      name
      species
      image
      gender
      episode {
        episode
        name
        air_date
      }
    }
  }
`;

export const SEARCH_CHARACTERS = gql`
  query GetCharacters($name: String!) {
    characters(filter: {name: $name}) {
      results {
        name
        image
        id
      }
    }
  }
`;
