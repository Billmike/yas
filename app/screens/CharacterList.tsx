import React from 'react';
import {
  View,
  StyleSheet,
  Text,
  KeyboardAvoidingView,
  FlatList,
  TouchableOpacity,
  Image,
  ActivityIndicator,
} from 'react-native';
import { useQuery } from '@apollo/client';
import Search from '../components/Search';
import { GET_CHARACTERS } from '../queries/index';

interface ICharacterList extends INavigationProps { }

type ItemProp = {
  item: {
    id: string;
    name: string;
    image: string;
  };
  handleItemPress: (id: string) => void;
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    paddingHorizontal: 20,
  },
  image: {
    height: 250,
    width: '100%',
  },
  list: {
    marginTop: 25,
    paddingBottom: 100,
  },
  name: {
    marginVertical: 15,
    textAlign: 'center',
  },
});

const Item = ({ item, handleItemPress }: ItemProp) => {
  return (
    <TouchableOpacity onPress={() => handleItemPress(item?.id)}>
      <Image
        source={{ uri: item?.image }}
        style={styles.image}
        resizeMode="contain"
      />
      <Text style={styles.name}>{item?.name}</Text>
    </TouchableOpacity>
  );
};

const CharacterList = ({ navigation }: ICharacterList) => {
  const { loading, data, fetchMore } = useQuery(GET_CHARACTERS, {
    fetchPolicy: 'cache-first',
    variables: {
      name: '',
    },
  });

  const handleItemPress = (id: string) => {
    navigation?.navigate('Details', { id });
  };

  const renderLoading = () => {
    if (loading) {
      return <ActivityIndicator size="large" />;
    }
    return null;
  };

  const handleFetchMore = () => {
    if (data?.characters.info.next) {
      if (fetchMore !== undefined) {
        return fetchMore({
          variables: {
            page: data?.characters?.info.next,
            name: '',
          },
        });
      }
    }
  };

  return (
    <View style={styles.container}>
      <KeyboardAvoidingView enabled>
        <Search />
        <FlatList
          data={data?.characters.results}
          renderItem={({ item }) => (
            <Item item={item} handleItemPress={handleItemPress} />
          )}
          keyExtractor={(item, index) => `${index}`}
          contentContainerStyle={styles.list}
          ListHeaderComponent={renderLoading}
          onEndReached={handleFetchMore}
        />
      </KeyboardAvoidingView>
    </View>
  );
};

export default CharacterList;
