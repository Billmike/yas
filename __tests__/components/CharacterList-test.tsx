import React from 'react';
import { render } from '@testing-library/react-native';
import CharacterList from '../../app/screens/CharacterList';
import { MockedProvider } from '@apollo/client/testing';
import { GET_CHARACTERS } from '../../app/queries/index';

const mocks = [
  {
    request: {
      query: GET_CHARACTERS,
      variables: {
        name: '',
      },
    },
    result: {
      data: {
        characters: {
          info: { next: 2 },
          results: [{ name: 'Rick', id: '1', image: 'dummyimage' }, { name: 'Rick', id: '1', image: 'dummyimage' }, { name: 'Rick', id: '1', image: 'dummyimage' }, { name: 'Rick', id: '1', image: 'dummyimage' }, { name: 'Rick', id: '1', image: 'dummyimage' }],
        },
      },
    },
  },
];

test('should render without error', () => {
  const { toJSON } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <CharacterList />
    </MockedProvider>,
  );
  expect(toJSON()).toMatchSnapshot();
});
