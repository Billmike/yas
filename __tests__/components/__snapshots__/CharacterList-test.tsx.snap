// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without error 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingHorizontal": 20,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "black",
          "borderRadius": 8,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "marginTop": 20,
          "paddingHorizontal": 15,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 25,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="search-icon"
      >
        î¢¶
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Find a character"
        rejectResponderTermination={true}
        style={
          Object {
            "flex": 1,
            "paddingLeft": 10,
          }
        }
        testID="search"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <RCTScrollView
      ListHeaderComponent={[Function]}
      contentContainerStyle={
        Object {
          "marginTop": 25,
          "paddingBottom": 100,
        }
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <ActivityIndicator
            animating={true}
            color="#999999"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
